### DISCUSSION TITLE
Closing sql.Rows ASAP on context.timeout

### DISCUSSION THREAD

**Discussion by sindy-senorita** [godror/godror#167] [Q&A] [answered]
Hi, I just realized that we have this notice

>  when the Context is canceled/timeouted. You should always call rows.Close ASAP, but if you experience random Breaks, remember this warning!

and think it might be related to the problem I have. I'm currently using context.Timeout for `db.QueryContext(ctx)`, and it stucks when timeout hits. With following sample code

```
// a very short timeout just to test the timeout handling in my app
ctx, cancel := context.WithTimeout(context.Background(), 600 * time.Millisecond) 

rows, err := db.QueryContext(ctx, "simple select query on a view object")
if err != nil {
  return err
}
defer func() {
  _ = rows.Close()
}
for rows.Next() {
  if ctx.Err() == context.DeadlineExceeded {
    return ctx.Err()
  }
  var resultModel ResultModel
  err = rows.Scan(&resultModel.field)
  if err != nil {
    return err
  }
  // other simple calculations and append
}
```

But it is still stucks at `db.QueryContext` statement. Printing db.Stats() every second shows 1 connection in use for 5 min long (I assume 5 min is the default pool timeout?)
```{MaxOpenConnections:10 OpenConnections:4 InUse:1 Idle:3 WaitCount:0 WaitDuration:0s MaxIdleClosed:0 MaxIdleTimeClosed:0 MaxLifetimeClosed:0}```
Then, immediately followed by `db.QueryContext` finally finish with `err: context deadline exceeded`; then connection InUse drops to 0.

Looks like the connection is stuck. I suspect that my method of closing rows ASAP is incorrect. Or could this be something else?

Thanks 🙌

**Comment by tgulacsi** [comment]
Where is it stuck, exactly? (Put println everywhere)

Where do you call cancel() ?

**Reply by sindy-senorita** [reply]
Oh, I forget to put it in the sample code
it is right after context creation
```
ctx, cancel := context.WithTimeout(context.Background(), 600 * time.Millisecond) 
defer cancel()

// then start working with db
```

**Reply by sindy-senorita** [reply]
and yes, I put println everywhere. Let me rewrite the sample code (Sorry I can't provide real code bcs it is still confidential 😅 )
```
// a very short timeout just to test the timeout handling in my app
ctx, cancel := context.WithTimeout(context.Background(), 600 * time.Millisecond) 
defer cancel()

fmt.Printf("start execute. context: %+v. now: %+v. stats: %+v\n", ctx, time.Now(), db.Stats())
rows, err := db.QueryContext(ctx, "simple select query on a view object")
fmt.Printf("done execute. context: %+v. now: %+v. stats: %+v\n", ctx, time.Now(), db.Stats())

if err != nil {
  return err
}
defer func() {
  _ = rows.Close()
}
for rows.Next() {
  if ctx.Err() == context.DeadlineExceeded {
    return ctx.Err()
  }
  var resultModel ResultModel
  err = rows.Scan(&resultModel.field)
  if err != nil {
    return err
  }
  // other simple calculations and append
}
```

last thing to print before stuck is the "start execute". the context is still alive then, and db stats shows no connection in use. After that, the inuse connection stays 1 for 5 minutes. Then print "done execute", inuse connection 0.
The query is a really simple select some data which successful if I put the timeout higher

**Comment by tgulacsi** [comment]
PTAL `z_test.go`, TestIssue100 - it's almost the same as what you've posted here.
Try it with your database, and try your test with the default test DB (ADW)!

**Reply by sindy-senorita** [reply]
Ok, tested. Copy-pasted the code with some modifications, but it still happens. The modifications:
- use plain select instead of create function ( I'm working with 3rd party system and I only have read-only privilege)
- reduce the timeout to 500ms to ensure I get the timeout

More observation on this shows that this might be intermittent. in a single call, 1st call is usually ok (`QueryContext` returns immediately after ctx timeout), 2nd sometimes hangs, 3rd hangs most of the time
When I wind up 4 goroutine that calls it, 1 out of 3 hangs most of the time. 
Could there be some racing/intermittent issue? Is there any more info that I can provide to help you identify it? Thanks for this.

Btw, if there is no feasible fix, my app can still be usable with the 5-minute freeze (this is a background job of a sort). But I would like to configure the 5 minutes wait into 1 minute. 
Based on [this](https://godror.github.io/godror/doc/connection.html), I tried to modify it from the connection string `oracle://username:password@host:port/dbname?connect_timeout=5&poolWaitTimeout=1m`. But, it doesn't seem to work (it still takes 5 minute for the inuse connection to be released)

**Reply by sindy-senorita** [reply]
oh yeah btw, `connect_timeout` value there doesn't seem to show any effect. What is it for? 
I'm sorry if I haven't been good at reading oracle doc (it is just too many versions and types out there). [This one says it is for the TCP call](https://docs.oracle.com/en/database/oracle/oracle-database/12.2/netrf/local-naming-parameters-in-tnsnames-ora-file.html#:~:text=TRANSPORT_CONNECT_TIMEOUT-,6.11.1%20CONNECT_TIMEOUT,-Purpose). So, I did an experiment. Turn off the internet connection, trigger the query execution. Blocks when doing `QueryContext`, nothing happens, connection inuse 1. The rest is still the same. It is waiting for 5-minute before unblocking the `QueryContext`. 


`SELECT * FROM V$VERSION;` on the database has
```
BANNER                                                                      |CON_ID|
----------------------------------------------------------------------------+------+
Oracle Database 12c Enterprise Edition Release 12.2.0.1.0 - 64bit Production|     0|
PL/SQL Release 12.2.0.1.0 - Production                                      |     0|
CORE 12.2.0.1.0 Production                                                  |     0|
TNS for Linux: Version 12.2.0.1.0 - Production                              |     0|
NLSRTL Version 12.2.0.1.0 - Production                                      |     0|
```

**Reply by tgulacsi** [reply]
> Ok, tested. Copy-pasted the code with some modifications, but it still happens. The modifications:
> 
>     * use plain select instead of create function ( I'm working with 3rd party system and I only have read-only privilege)
> 
>     * reduce the timeout to 500ms to ensure I get the timeout
> 
> 

Please tweak rows.go, line 292 - I don't really like the Context stored in statement.
Where is the deadlock/timeout (set godror.Log or DPI_DEBUG_LEVEL=92) ?

**Reply by tgulacsi** [reply]
With ShortTimeout, something interesting can be achieved, with properly sized timeout:

```
$ DPI_DEBUG_LEVEL=92 VERBOSE=1 go test -run=ShortTimeout -v
ODPI [1960058] 2021-08-07 14:13:50.322: ODPI-C 4.2.1
ODPI [1960058] 2021-08-07 14:13:50.322: debugging messages initialized at level 92
eDSN:
Using default database for tests:  /home/tgulacsi/src/github.com/godror/godror/contrib/free.db/env.sh
export TNS_ADMIN="/home/tgulacsi/src/github.com/godror/godror/contrib/free.db"
export GODROR_TEST_DSN="user=test password=r97oUPimsmTOIcBaeeDF connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-3904033253 connectionClass=TestClassName enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= newPassword= noTimezoneCheck=0\npoolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s prelim=0 standaloneConnection=1\nsysasm=0 sysdba=0 sysoper=0 timezone=UTC"
msg=connect poolParams="poolIncrement=2 poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s\npoolSessionTimeout=1m0s poolWaitTimeout=5s" connParams= common="user=test password=\"SECRET-lc79Key-hlk=\" connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-3904033253 enableEvents=1 timezone=UTC"
msg=dpiContext_createWithParams params="&{defaultDriverName:0x13b9aa0 defaultEncoding:0x13b9a80 loadErrorUrl:<nil> oracleClientLibDir:<nil> oracleClientConfigDir:0x13b9eb0}"
ODPI [1960058] 2021-08-07 14:13:50.326: fn start dpiContext_createWithParams
ODPI [1960058] 2021-08-07 14:13:50.326: Context Parameters:
ODPI [1960058] 2021-08-07 14:13:50.326:     Oracle Client Config Dir: /tmp/godror_drv_test-3904033253
ODPI [1960058] 2021-08-07 14:13:50.326: Environment Variables:
ODPI [1960058] 2021-08-07 14:13:50.326:     ORACLE_HOME => "/usr/lib/oracle/21/client64"
ODPI [1960058] 2021-08-07 14:13:50.326:     TNS_ADMIN => "/tmp/godror_drv_test-3904033253"
ODPI [1960058] 2021-08-07 14:13:50.326: check module directory
ODPI [1960058] 2021-08-07 14:13:50.326: module name is /tmp/go-build1929854413/b001/godror.test
ODPI [1960058] 2021-08-07 14:13:50.326: load in dir /tmp/go-build1929854413/b001
ODPI [1960058] 2021-08-07 14:13:50.326: load with name /tmp/go-build1929854413/b001/libclntsh.so
ODPI [1960058] 2021-08-07 14:13:50.326: load by OS failure: /tmp/go-build1929854413/b001/libclntsh.so: cannot open shared object file: No such file or directory
ODPI [1960058] 2021-08-07 14:13:50.326: load with OS search heuristics
ODPI [1960058] 2021-08-07 14:13:50.326: load with name libclntsh.so
ODPI [1960058] 2021-08-07 14:13:50.332: load by OS successful
ODPI [1960058] 2021-08-07 14:13:50.332: validating loaded library
ODPI [1960058] 2021-08-07 14:13:50.336: fn end dpiContext_createWithParams -> 0
ODPI [1960058] 2021-08-07 14:13:50.336: fn start dpiContext_getClientVersion(0x1437100)
ODPI [1960058] 2021-08-07 14:13:50.336: fn end dpiContext_getClientVersion(0x1437100) -> 0
msg=acquireConn pool=<nil> connParams="user=test password=\"SECRET-lc79Key-hlk=\" connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-3904033253 enableEvents=1 timezone=UTC "
ODPI [1960058] 2021-08-07 14:13:50.336: fn start dpiContext_initCommonCreateParams(0x1437100)
ODPI [1960058] 2021-08-07 14:13:50.336: fn end dpiContext_initCommonCreateParams(0x1437100) -> 0
ODPI [1960058] 2021-08-07 14:13:50.336: fn start dpiContext_initConnCreateParams(0x1437100)
ODPI [1960058] 2021-08-07 14:13:50.336: fn end dpiContext_initConnCreateParams(0x1437100) -> 0
ODPI [1960058] 2021-08-07 14:13:50.336: fn start dpiConn_create(0x1437100)
ODPI [1960058] 2021-08-07 14:13:51.064: OCI error ORA-28002: the password will expire within 28 days (dpiConn_create / begin session)
ODPI [1960058] 2021-08-07 14:13:51.064: fn end dpiConn_create(0x1437100) -> 0
msg="init connection" conn="&user=test password=\"SECRET-lc79Key-hlk=\" connectString=db201911301540_high\nconfigDir=/tmp/godror_drv_test-3904033253 connectionClass= enableEvents=1\nexternalAuth=0 heterogeneousPool=0 libDir= noTimezoneCheck=0 poolIncrement=0\npoolMaxSessions=0 poolMinSessions=0 poolSessionMaxLifetime=0s poolSessionTimeout=0s\npoolWaitTimeout=0s prelim=0 standaloneConnection=0 sysasm=0 sysdba=0 sysoper=0\ntimezone=UTC&serverVersion=0.0.0.0.0&tzOffSecs=0&new=true" onInit=<nil>
msg=initTZ tzValid=false paramsTZ=UTC
msg=setCallTimeout ms=29260
ODPI [1960058] 2021-08-07 14:13:51.064: fn start dpiConn_setCallTimeout(0x1437f20)
ODPI [1960058] 2021-08-07 14:13:51.064: fn end dpiConn_setCallTimeout(0x1437f20) -> 0
ODPI [1960058] 2021-08-07 14:13:51.064: fn start dpiConn_ping(0x1437f20)
ODPI [1960058] 2021-08-07 14:13:51.100: fn end dpiConn_ping(0x1437f20) -> 0
msg=setCallTimeout ms=0
ODPI [1960061] 2021-08-07 14:13:51.101: fn start dpiConn_setCallTimeout(0x1437f20)
ODPI [1960061] 2021-08-07 14:13:51.102: fn end dpiConn_setCallTimeout(0x1437f20) -> 0
msg=IsValid connOK=true released=false pooled=false tzOK=true
# user=test password="SECRET-lc79Key-hlk=" connectString=db201911301540_high
configDir=/tmp/godror_drv_test-3904033253 connectionClass=TestClassName enableEvents=1
externalAuth=0 heterogeneousPool=0 libDir= noTimezoneCheck=0 poolIncrement=2
poolMaxSessions=16 poolMinSessions=2 poolSessionMaxLifetime=5m0s poolSessionTimeout=1m0s
poolWaitTimeout=5s prelim=0 standaloneConnection=1 sysasm=0 sysdba=0 sysoper=0
timezone=UTC
Version: v0.25.4
Client: 21.1.0.0.0 Timezone: Local
ODPI [1960061] 2021-08-07 14:13:51.103: fn start dpiConn_getServerVersion(0x1437f20)
ODPI [1960061] 2021-08-07 14:13:51.135: fn end dpiConn_getServerVersion(0x1437f20) -> 0
Server: 19.12.0.1.0 [Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production; Version 19.12.0.1.0] Timezone: UTC
msg=IsValid connOK=true released=false pooled=false tzOK=true
=== RUN   TestShortTimeout
ODPI [1960061] 2021-08-07 14:13:51.136: fn start dpiConn_setModule(0x1437f20)
ODPI [1960061] 2021-08-07 14:13:51.136: fn end dpiConn_setModule(0x1437f20) -> 0
ODPI [1960061] 2021-08-07 14:13:51.136: fn start dpiConn_prepareStmt(0x1437f20)
ODPI [1960061] 2021-08-07 14:13:51.136: SQL SELECT * FROM all_objects ORDER BY DBMS_RANDOM.value
ODPI [1960061] 2021-08-07 14:13:51.137: fn end dpiConn_prepareStmt(0x1437f20) -> 0
ODPI [1960061] 2021-08-07 14:13:51.137: fn start dpiStmt_getInfo(0x7fa6a0047560)
ODPI [1960061] 2021-08-07 14:13:51.137: fn end dpiStmt_getInfo(0x7fa6a0047560) -> 0
ODPI [1960061] 2021-08-07 14:13:51.137: fn start dpiStmt_getBindCount(0x7fa6a0047560)
ODPI [1960061] 2021-08-07 14:13:51.137: fn end dpiStmt_getBindCount(0x7fa6a0047560) -> 0
enter=bindVars st=0xc000254000 args=[]
doManyCount=1 arrLen=-1 doExecMany=false minArrLen=maxArrLen
ODPI [1960061] 2021-08-07 14:13:51.137: fn start dpiStmt_setFetchArraySize(0x7fa6a0047560)
ODPI [1960061] 2021-08-07 14:13:51.137: fn end dpiStmt_setFetchArraySize(0x7fa6a0047560) -> 0
ODPI [1960061] 2021-08-07 14:13:51.137: fn start dpiStmt_setPrefetchRows(0x7fa6a0047560)
ODPI [1960061] 2021-08-07 14:13:51.137: fn end dpiStmt_setPrefetchRows(0x7fa6a0047560) -> 0
ODPI [1960061] 2021-08-07 14:13:51.137: fn start dpiStmt_execute(0x7fa6a0047560)
msg="BREAK context statement" error="context deadline exceeded"
msg=Break dpiConn="&{typeDef:0x97ab40 checkInt:1239179276 refCount:2 env:0x146b7f0 pool:<nil> handle:0x149ca30 serverHandle:0x149db58 sessionHandle:0x1533b48 shardingKey:<nil> superShardingKey:<nil> transactionHandle:<nil> releaseString:0x7fa6a00474b0 releaseStringLength:90 rawTDO:<nil> versionInfo:{versionNum:19 releaseNum:12 updateNum:0 portReleaseNum:1 portUpdateNum:0 fullVersionNum:1912000100} commitMode:0 charsetId:873 openStmts:0x14360f0 openLobs:0x146b940 objects:0x146b9e0 externalHandle:0 deadSession:0 standalone:1 creating:0 closing:0 _:[0 0 0 0]}"
ODPI [1960058] 2021-08-07 14:13:51.637: fn start dpiConn_breakExecution(0x1437f20)
ODPI [1960058] 2021-08-07 14:13:51.638: fn end dpiConn_breakExecution(0x1437f20) -> 0







ODPI [1960061] 2021-08-07 14:13:58.853: OCI error ORA-04045: errors during recompilation/revalidation of SYS.DBMS_WORKLOAD_REPLAY_I
ORA-00900: invalid SQL statement
ORA-01013: user requested cancel of current operation (dpiStmt_execute / execute)
ODPI [1960061] 2021-08-07 14:13:58.853: fn end dpiStmt_execute(0x7fa6a0047560) -> -1
msg=statement.closeNotLocking st=0xc000254000 refCount=1
ODPI [1960061] 2021-08-07 14:13:58.853: fn start dpiStmt_release(0x7fa6a0047560)
ODPI [1960061] 2021-08-07 14:13:58.853: fn end dpiStmt_release(0x7fa6a0047560) -> 0
msg=IsValid connOK=true released=false pooled=false tzOK=true
    z_test.go:3759: <nil> context deadline exceeded
    z_test.go:3763: context deadline exceeded
--- PASS: TestShortTimeout (7.72s)
PASS
ODPI [1960061] 2021-08-07 14:13:58.854: fn start dpiConn_release(0x1437f20)
ODPI [1960061] 2021-08-07 14:13:58.892: fn end dpiConn_release(0x1437f20) -> 0
ODPI [1960061] 2021-08-07 14:13:58.892: fn start dpiContext_destroy(0x1437100)
ODPI [1960061] 2021-08-07 14:13:58.892: fn end dpiContext_destroy(0x1437100) -> 0
ok      github.com/godror/godror        8.575s
```

So, it seems that breakExecution is called, but waits a while.
Maybe it's normal, I don't know.

**Reply by tgulacsi** [reply]
Using dpiConn_setCallTimeout helps the faster recover (no Break needed), 
but borks everything else by rendering the connection unusable.

**Reply by sindy-senorita** [reply]
Hi @tgulacsi . Thanks. increasing the debug level helps me understand the scenario better now.
It is just like your test, when break(s) happen, connections that are still executing `dpiStmt_fetchRow` or `dpiStmt_execute` stuck for >= 5minutes. But, in my case, it gets `OCI error ORA-03113: end-of-file on communication channel`.

I have repeated the run > 5 times and the result is consistent.

[here is the complete log in gist ](https://gist.github.com/kerocen/776cd8e1fc8a70b41382285f0e429fb4)

Things to be observed from the log, I've started 4 select calls, concurrently to 4 different views (namely VIEW_1 - VIEW_4). 

- on view 1 (00338) finished ok (read all rows ok, and to the point of `fn end dpiConn_release(0x7f22d004b250) -> 0`)
- on view 2 (00334) and view 4 (00337), conn are stuck when break happens. On view 2 has just started execution (`fn start dpiStmt_execute(0x2389eb0)`) . On view 4, stmt has started row fetch (`fn start dpiStmt_fetchRows(0x7f22dc129120)`)
- on view 3 (I assume it is 00341) deadline exceeded when it is still grabbing the connection `fn start dpiConn_create(0x7f22dc07bf30)`. But, unlike view 2 and 4, this one closed successfully after the breaks
- there is two other connection: 00340 and 00336 that initiate the breaks

Does it help?
What do you mean by using dpiConn_setCallTimeout?

**Reply by tgulacsi** [reply]
@kerocen  can you try the "timeout" branch? It adds a OCI call timeout, that may help.

**Comment by sudarshan12s** [comment]
@kerocen,
Can you share details of instance client you are using? complete program
with goroutines used, if possible
I remember older mac instant client versions had issues
with poolWaitTimeout . new versions of mac client
<https://www.oracle.com/database/technologies/instant-client/macos-intel-x86-downloads.html>

**Reply by sindy-senorita** [reply]
Hi @sudarshan12s . Thanks
I'm using mac but the app is run on docker container (Linux) image golang 1.15.
We are using the lib from https://www.oracle.com/database/technologies/instant-client/linux-x86-64-downloads.html
and already the latest (21.1.0.0..)

**Comment by sudarshan12s** [comment]
I could not simulate exact behaviour.
@Kerocen, it looks the break/reset operation which depend on OOB (out of
band ) or inband support was not successful in your case resulting in
ORA-3113.
does it come with inband (sqlnet.ora , DISABLE_OOB=on)  too

**Reply by sindy-senorita** [reply]
hi @sudarshan12s , which one are you trying to simulate? the break or the poolWaitTimeout?

> does it come with inband (sqlnet.ora , DISABLE_OOB=on)  too

Sorry, I'm not sure if I understand the statement. How can I check this, is this from the server configuration?

**Comment by sudarshan12s** [comment, marked as answer]
Hi @kerocen,

poolWaitTimeout is applied during getconnection. looks this is fine in your
case for thread 00341.
break is applicable for connections which are doing active query like
stmtexec/stmtfetch which was failing for threads 00337 an 00334.
a break is issued oob (out of band) by default from client. it can be
turned off to use inband. This can be done by changing sqlnet.ora only at
client side to
DISABLE_OOB=on

By default client issues OOB breaks  .  If firewall blocks OOB,
client switches to inband with latest DB versions.
In your case (DB 12.c) , Can you check firewall settings,  if they block
OOB.  In that case, does changing sqlnet.ora to use inband help ?

**Reply by sindy-senorita** [reply]
I think that helps! But I'll run some test code around for a total sure.
In meantime, do you know if there is any caveat/drawback of using inbound vs outbound breaks? @sudarshan12s

**Reply by sudarshan12s** [reply]
@kerocen, I was just asking for confirmation. inband notifications are  generally slow . oob is faster .

**Reply by sindy-senorita** [reply]
Yes, blocked OOB was the cause. It is working now, ctx canceled/timeout returns without any delay. Thanks @sudarshan12s , @tgulacsi
